File: packages/typescript/src/cpfValidator.ts
const defaultErrorMsg: string[] = [
	"CPF invalid",
	"CPF must have 11 numerical digits",
	"CPF is not valid",
	"Unknown error",
];

/**
 * Validates a Brazilian CPF number for correctness.
 *
 * The CPF (Cadastro de Pessoas Físicas) is a Brazilian tax identification number.
 * It consists of 11 digits in the format XXX.XXX.XXX-XX. This function checks the
 * validity of a CPF number using its calculation algorithm.
 *
 * @param cpf The CPF number as a string.
 * @param errorMsg An optional array of custom error messages.
 * @example cpfIsValid('123.456.789-09');
 * @example cpfIsValid('12345678909');
 * @example cpfIsValid('12345678909', ['Custom invalid CPF message','Custom length error']);
 * @returns An object with 'isValid' (boolean) and 'errorMsg' (string) properties.
 */
function cpfIsValid(
	cpf: string,
	errorMsg: (string | null)[] | null = defaultErrorMsg,
): {
	isValid: boolean;
	errorMsg: string | null;
} {
	if (typeof cpf !== "string") {
		throw new TypeError("The input should be a string.");
	}
	if (errorMsg) {
		if (!Array.isArray(errorMsg)) throw new TypeError("Must be an Array");
		for (let index: number = 0; index < errorMsg.length; index += 1) {
			if (errorMsg[index] != null && typeof errorMsg[index] !== "string") {
				throw new TypeError(
					"All values within the array must be strings or null/undefined.",
				);
			}
		}
	}

	function getErrorMessage(index: number): string {
		const errorMessage: string | null = errorMsg ? errorMsg[index] : null;
		return errorMessage != null ? errorMessage : defaultErrorMsg[index];
	}

	try {
		if (!cpf) {
			return {
				isValid: false,
				errorMsg: getErrorMessage(0),
			};
		}

		const cpfClean: string = cpf.replace(/\D+/g, "");
		
		if (/^(\d)\1{10}$/.test(cpfLimpo)) {
			return {
				isValid: false,
				errorMsg: getErrorMessage(2),
			};
		}

		if (cpfLimpo.length !== 11) {
			return {
				isValid: false,
				errorMsg: getErrorMessage(1),
			};
		}

		const cpfArray: number[] = cpfClean.split('').map(Number);
		const validator = (sum: number) => (sum % 11 < 2 ? 0 : 11 - (sum % 11));
		const sum1 = cpfArray.slice(0, 9).reduce((acc, val, i) => acc + val * (10 - i), 0);
		const sum2 = cpfArray.slice(0, 10).reduce((acc, val, i) => acc + val * (11 - i), 0);

		if (cpfArray[9] === validator(sum1) && cpfArray[10] === validator(sum2)) {
			return {
				isValid: true,
				errorMsg: null,
			};
		}
		return {
			isValid: false,
			errorMsg: getErrorMessage(2),
		};
	} catch (err) {
		return {
			isValid: false,
			errorMsg: getErrorMessage(3),
		};
	}
}

export default cpfIsValid;

File: packages/c/src/cpfValidator.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <stdbool.h>

bool isAllSameDigit(const char *str) {
    char firstChar = str[0];
    for (int i = 1; i < 11; i++) {
        if (str[i] != firstChar) {
            return false;
        }
    }
    return true;
}

bool cpfIsValid(const char *cpf) {
    if (cpf == NULL) {
        fprintf(stderr, "CPF cannot be null or empty\n");
        return false;
    }

    char cpfClean[12] = {0};
    int index = 0;

    for (int i = 0; i < strlen(cpf); i++) {
        if (isdigit(cpf[i])) {
            cpfClean[index++] = cpf[i];
        }
    }

    if (strlen(cpfClean) != 11) {
        return false;
    }

    if (isAllSameDigit(cpfClean)) {
        return false;
    }

    int cpfArray[11];
    
    for (int i = 0; i < 11; i++) {
        cpfArray[i] = cpfClean[i] - '0';
    }

    int sum1 = 0, sum2 = 0;

    for (int i = 0; i < 9; i++) {
        sum1 += cpfArray[i] * (10 - i);
    }

    for (int i = 0; i < 10; i++) {
        sum2 += cpfArray[i] * (11 - i);
    }

    int validator1 = sum1 % 11 < 2 ? 0 : 11 - (sum1 % 11);
    int validator2 = sum2 % 11 < 2 ? 0 : 11 - (sum2 % 11);

    return cpfArray[9] == validator1 && cpfArray[10] == validator2;
}

int main(int argc, char *argv[]) {
    if (argc > 2) {
        fprintf(stderr, "Usage: %s [cpf]\n", argv[0]);
        return 1;
    }

    const char *cpf = argv[1];
    
    if (cpfIsValid(cpf)) {
        printf("CPF is valid.\n");
    } else {
        printf("CPF is invalid.\n");
    }
    return 0;
}
